NEP: add NEP 56 on array API standard support in main namespace

rgommers opened the pull request:
For NumPy 2.0. xref gh-25076 as the main tracking issue for this topic.

rgommers committed:
NEP: add NEP 56 on array API standard support in main namespace

For NumPy 2.0. xref gh-25076 as the main tracking issue for this topic.

[skip actions] [skip azp] [skip cirrus]

mhvk reviewed (commented):
Nice! A few small comments; the only one that needs some thought is the one about the meaning of "array scalar"

rgommers commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+rather than less consistent.
+
+We'll note that one remaining incompatibility will be that NumPy is returning
+array scalars rather than 0-D arrays in most cases where the standard, and
```
I don't think it does. E.g., from https://numpy.org/devdocs/reference/arrays.scalars.html#scalars: _"Array scalars have the same attributes and methods as `ndarrays`. This allows one to treat items of an array partly on the same footing as arrays, smoothing out rough edges that result when mixing scalar and array operations."_

There is no special name for 0-D arrays AFAIK.

mhvk replied:
Hah, learned something new!

lucascolley commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+
+Benefits:
+
+- It will remove the "having to make a choice between the NumPy API and the
+  Array API" issue for other libraries,
```
For consistency
```suggestion
- It will remove the "having to make a choice between the NumPy API and the
  array API" issue for other libraries,
```

lucascolley commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+- Expanding the size of the main namespace with about ~20 aliases (e.g.,
+  ``acos`` & co. with C99 names aliasing ``arccos`` & co.).
+
+Overall we believe that the benefits significantly outweigh the gains - and are
+permanent, while the costs are largely temporary. In particular, the benefits
```
Typo
```suggestion
Overall we believe that the benefits significantly outweigh the costs - and are
permanent, while the costs are largely temporary. In particular, the benefits
```

lucascolley commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+Overall we believe that the benefits significantly outweigh the gains - and are
+permanent, while the costs are largely temporary. In particular, the benefits
+to array libraries and compilers that want to achieve compatibility with NumPy
+are significant. And as a result, the long-term benefits for the PyData (or
+scientific Python) ecosystem as a whole - because of downstream libraries being
+able to support multiple array libraries as easily as possible - are
+significant too. The number of breaking changes needed is fairly limited, and
```
opinion: I'd use slightly weaker wording here, as reaching the "as easily as possible" point will require slightly more than just array API adoption.
```suggestion
are significant. And as a result, the long-term benefits for the PyData (or
scientific Python) ecosystem as a whole - because of downstream libraries being
able to support multiple array libraries much more easily - are
significant too. The number of breaking changes needed is fairly limited, and
```

lucascolley commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+
+In scope for this NEP are:
+
+- Changes to NumPy's Python API needed to support the 2022.12 version of the array API standard,
```
I think we should be explicit about whether the `linalg` and `fft` extensions are included here.

lucascolley commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+array API standard support should save them time. The design rules ([3]_), and
+in some cases new APIs like the ``unique_*`` ones, are easier to implement on
+GPU and for JIT compilers as a result of more predictable behavior.
+
```
Perhaps this is a bit too speculative or big-picture, but I think a comment about the potential impact on new (in the sense of not done before with any array library) functionality - only needing to have 1 array-agnostic implementation, rather than 5+ similar-but-incompatible implementations for each array library ecosystem - wouldn't be out of place while we are talking about consumer libraries.

rgommers replied:
If you don't mind, I'd like to leave this as is, since it's described in the array API standard language, in NEP 47, and in blog posts, talks, etc. - I'm running short on time for 2.0, and there's only so many times I can rewrite the same abstract point. NEP 47 has for example:

_For a longer discussion on the purpose of the array API standard we refer to the [Purpose and Scope section of the array API standard](https://data-apis.github.io/array-api/latest/purpose_and_scope.html) and the two blog posts announcing the formation of the Consortium [[1]](file:///home/rgommers/code/numpy/doc/neps/_build/html/nep-0047-array-api-standard.html#id6) and the release of the first draft version of the standard for community review [[2]](file:///home/rgommers/code/numpy/doc/neps/_build/html/nep-0047-array-api-standard.html#id7)._

lucascolley replied:
Sure thing!

lucascolley commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+1. Raising errors for consistency/strictness in some places where NumPy now
+   allows more flexible behavior,
+2. Dtypes of returned arrays for some element-wise functions and reductions,
+3. Numerical behavior for few tolerance keywords,
```
Not sure but I think a typo?
```suggestion
3. Numerical behavior for a few tolerance keywords,
```

lucascolley commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+Adapting to the changes
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Some part of the Array API has already been implemented as part of the general
+Python API cleanup for NumPy 2.0 (see NEP 52), such as:
```
```suggestion
Some parts of the array API have already been implemented as part of the general
Python API clean-up for NumPy 2.0 (see NEP 52), such as:
```

lucascolley commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+part of the array API standard nor APIs with some types of backwards
+incompatible changes discussed above.
+
+For an automated migration to Array API compatible codebase, a new rule is
+being implemented (see issue `ruff#8615 <https://github.com/astral-sh/ruff/issues/8615>`__
+and PR `ruff#8910 <https://github.com/astral-sh/ruff/pull/8910>`__).
```
```suggestion
For an automated migration to an array API compatible codebase, a new rule is
being implemented (see issue `ruff#8615 <https://github.com/astral-sh/ruff/issues/8615>`__
and PR `ruff#8910 <https://github.com/astral-sh/ruff/pull/8910>`__).
```

lucascolley commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+being implemented (see issue `ruff#8615 <https://github.com/astral-sh/ruff/issues/8615>`__
+and PR `ruff#8910 <https://github.com/astral-sh/ruff/pull/8910>`__).
+
+With both rules in place a downstream user should be able to update their
+project, to the extent that that is possible with automation, to a library
+agnostic codebase that can benefit from different array libraries and devices.
```
```suggestion
With both rules in place a downstream user should be able to update their
project, to the extent that is possible with automation, to a library
agnostic codebase that can benefit from different array libraries and devices.
```

lucascolley commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+project, to the extent that that is possible with automation, to a library
+agnostic codebase that can benefit from different array libraries and devices.
+
+Backwards incompatible changes that cannot be handled automatically (e.g., a
+change in ``rtol`` defaults for a linear algebra function), this will be
+handled the same way as any other backwards incompatible change in NumPy 2.0 -
+through documentation, release notes, and API migrations and deprecations over
+several releases.
```
```suggestion
Backwards incompatible changes that cannot be handled automatically (e.g., a
change in ``rtol`` defaults for a linear algebra function) will be handled
in the same way as any other backwards incompatible change in NumPy 2.0 -
through documentation, release notes, API migrations and deprecations over
several releases.
```

lucascolley commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+Detailed description
+--------------------
+
+In this section we'll focus on specific API additions and functionality that we
+would not consider introducing into NumPy if the standard would not exist and
+we wouldn't have to think/worry about its main goal of writing code that is
+portable across multiple array libraries and their supported features like GPUs
+and other hardware accelerators or JIT compilers.
```
This is quite a long sentence to have no punctuation - a suggestion but whatever you think reads best :)
```suggestion
In this section, we'll focus on specific API additions and functionality that we
would not consider introducing into NumPy if the standard did not exist and
we didn't have to think/worry about its main goal: writing code that is
portable across multiple array libraries and their supported features like GPUs
and other hardware accelerators or JIT compilers.
```

lucascolley commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+rationale. In this case, it is needed in order to match other libraries.
+The main set of aliases added is for trigonometric functions, where
+the array API standard chose to follow C99 and other libraries in using
+``acos``, ``asin`` etc. rather than ``arccos``, ``arcsin``, etc. NumPy usually
+also follows C99; it is not entirely clear why this naming choice was made many
+years ago.
```
minor
```suggestion
``acos``, ``asin`` etc. rather than ``arccos``, ``arcsin``, etc. NumPy also
usually follows C99; it is not entirely clear why this naming choice was made many
years ago.
```

lucascolley commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+
+The ``unique`` function, with ``return_index``, ``return_inverse``, and
+``return_counts`` arguments that influence the cardinality of the returned
+tuple, is replaced in the Array API by four respective functions:
```
for consistency
```suggestion
tuple, is replaced in the array API by four respective functions:
```

lucascolley commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+``np.bool`` addition
+^^^^^^^^^^^^^^^^^^^^
+
+One of the aliases that used to live in NumPy but got removed was ``np.bool``.
+To comply with the Array API it got reintroduced with a different meaning, as
```
for consistency
```suggestion
To comply with the array API it got reintroduced with a different meaning, as
```

lucascolley replied:
grammar nit, feel free to ignore
```suggestion
One of the aliases that used to live in NumPy but was removed is ``np.bool``.
To comply with the Array API it was reintroduced with a different meaning, as
```

lucascolley commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+To comply with the Array API it got reintroduced with a different meaning, as
+now it points to NumPy's bool instead of a Python builtin. This change is a
+good idea and we were planning to make it anyway, because ``bool`` is a nicer
+name than ``bool_``. However, we may not have scheduled that reintroduction of
+the name for 2.0 perhaps if it had not been part of the array API standard.
```
nit: no need for both 'may' and 'perhaps'
```suggestion
name than ``bool_``. However, we may not have scheduled that reintroduction of
the name for 2.0 if it had not been part of the array API standard.
```

lucascolley commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+
+The PR that was merged as one of the first contained a new CI job that adds the
+`array-api-tests <https://github.com/data-apis/array-api-tests>`__ test suite.
+This way we had a better control over which batch of functions/aliases is being
+added each time, and be sure that the implementation conforms with the Array
+API standard (see `gh-25167 <https://github.com/numpy/numpy/pull/25167>`__).
```
```suggestion
This way we had better control over which batch of functions/aliases were being
added each time, and could be sure that the implementations conformed to the array
API standard (see `gh-25167 <https://github.com/numpy/numpy/pull/25167>`__).
```

lucascolley reviewed (commented):
Great to see, nice work! A few comments but mostly minor editorial bits.

ngoldbaum commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+
+The ``copy`` keyword in ``astype`` will stick to its current meaning, because
+"never copy" when asking for a cast to a different dtype doesn't quite make
+sense.
```
xref https://github.com/numpy/numpy/pull/25168#issuecomment-1857878408

It would be nice to get a little bit more discussion about how downstream should adapt to this change. Just replace `np.array(..., copy=False)` with `np.asarray()`? If so, do we communicate that in the `ValueError` that gets generated?

seberg replied:
Yes, this is a at least almost a large impact change:  It should have its own sub section in the backwards compatibility section.  (It is currently completely missing there).

rgommers replied:
I assume you're commenting on `asarray`, not `astype`. I'd say the migration guide should say: do nothing (because that's more likely than not what the author intended - remember we have a semantic ambiguity here right now), and if you get an exception, inspect the code and if you indeed intended "only if needed", change it to `copy=None`.

I also don't think it's a large change. The default is do nothing, and in some cases the exception message will then tell you what to change.

seberg replied:
That is fair, although I still think it would be good to explicitly mention the BC change of having to replace `array(..., copy=False)` with `asarray(...)`.  You can point out that it isn't super common because `asarray()` makes more sense and that the it is turns into an easy to fix error.

I am actually not convinced that it's more likely than not what the author intended (unless you found that by code search).  Many cases are probably just better served by `asarray()` but tried to micro-optimize it away because it had extra overhead until a few years ago.
Also there is the caveat that if we go e.g. via `__array__()`, `copy=False` will fail even when it shouldn't conceptually.

mhvk replied:
I've used `np.array(..., copy=False, subok=True)` quite a lot, when indeed it was faster than `np.asanyarray`; I don't think I have ever had a cases where I'd want an error if the input was not an array...

That said, I agree with the logic of `False/True/None` and don't mind to change my code/astropy (though I wish `copy=None` was backward compatible).

mhvk replied:
> Also there is the caveat that if we go e.g. via `__array__()`, `copy=False` will fail even when it shouldn't conceptually.

This is more problematic, indeed.

rgommers replied:
Quick comments while I'm sorting through the text and gh-25168:
1. there's a bit of a circular thing going on here, where most of the problem seems to be `array` (a non-recommended / don't care function), and the only reason we even want to update `array` is for consistency with `asarray` (the important function here)
2. I'm not 100% sure I understand the `__array__` comment. I haven't thought about that one much before, but I think you mean something here like *"we cannot know if a call to `__array__` is going to cause this non-numpy object to copy data, so we must not call `__array__` if `copy=False`"?* If so, I'm not sure that's necessarily a long-term problem; a logical solution would be to add `copy=None` keyword to `__array__` so that we can know. If numpy cannot know it preferably shouldn't be guessing.

jakevdp commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+- CuPy's `docs on array API support <https://docs.cupy.dev/en/stable/reference/array_api.html>`__
+  and `PRs labelled with array-api <https://github.com/cupy/cupy/pulls?q=is%3Aopen+is%3Apr+label%3Aarray-api>`__.
+- JAX: enhancement proposal `Scope of JAX NumPy & SciPy Wrappers <https://jax.readthedocs.io/en/latest/jep/18137-numpy-scipy-scope.html#axis-2-array-api-alignment>`__
+  and `PR with initial implementation <https://github.com/google/jax/pull/16099>`__.
```
A more up-to-date reference is JAX's tracking issue for array API support: https://github.com/google/jax/issues/18353

seberg commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+- ``argsort`` and ``sort`` will gain a ``stable`` keyword argument in addition
+  to ``kind``, and the default will become ``stable=True``.
+- The ``ddof`` keyword in ``std`` and ``var`` changes its name to
+  ``correction``.
```
Do we actually have concrete plans to deprecate it?  IF yes, say that it will be a slow deprecation, as that is important to judge backcompat impact.
I.e. "we plan to deprecate the `ddof` name eventually".

rgommers replied:
I don't think we have plans to deprecate `ddof`. At least I didn't consider it yet, and I'm not sure it's necessary.

seberg reviewed (commented):
Some (partially opinionated) comments.  Overall looks nice.

I would prefer somewhat different emphasis at points (as you can probably read out of the comments), but that is opinions.

However, the backwards compatibility section needs to be much clearer about the larger changes/harder to grasp changes.
That might be a small note in the end, the main issue is that we need to try it before I can voice a real opinion.  For now I am worried by the change.

lucascolley commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+Benefits:
+
+- It will remove the "having to make a choice between the NumPy API and the
+  Array API" issue for other libraries,
```
I believe it is meant to be none of those options. Rather, for other _array-provider_ libraries, like CuPy and PyTorch, when they are  looking to implement new functionality, they have often followed NumPy, diverging only where they see benefits (CuPy follows NumPy more closely of course).

Now that we have the array API (still in early stages), these array-provider libraries don't really know whether following the NumPy API or the array API will be better for them (are differences going to persist? which will allow for better interoperability? which will make backwards compatibility easier?).

If NumPy conforms to the array API, these conflicts will disappear.

rgommers replied:
I meant both for producers and consumers, so roughly @seberg's second and third bullets and what Lucas said. But his first bullet is also a benefit. So all of these are good, except for `import torch as np` - that's probably not something we should encourage:)

Will try to rephrase.

leofang commented:
cc:@kmaehashi for vis

rgommers commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+has an integer dtype. It also includes dtype changes: ``sum`` and ``prod``
+always upcast lower-precision floating-point dtypes to ``float64`` when
+``dtype=None`` (this upcasting is already done for inputs with lower-precision
+integer dtypes). 
```
Yes, I think this is the one change I wrote down and then intended to revisit, and then forgot about. We haven't implemented it, and I don't think there's a real reason for doing so. I believe this is better to leave untouched indeed, and open an issue on the array API repo that this upcasting is allowed (or recommended), but not mandated.

rgommers replied:
I opened https://github.com/data-apis/array-api/issues/731 to reconsider this design choice in the standard. And I added a new section "_Parts of the standard that are not adopted_". That addresses this point I believe.

rgommers commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+tuple, is replaced in the Array API by four respective functions:
+``unique_all``, ``unique_counts``, ``unique_inverse``, and ``unique_values``.
+This new API helps to better anticipate the return object and support clearer
+typing.
```
No, static typing has problems with this too. If you look at the many overloads for `unique`, you'll see that they only work for `Literal[True]`/`Literal[False]`, anything else will not work for Mypy & co.

Adopting your suggestion to mention tools like Numba (and Mypy).

rgommers committed:
DOC: rework for review comments: textual suggestions and easy changes

[skip actions] [skip cirrus] [skip azp]

rgommers commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+
+A related note on terminology: "stacking" is a confusing and fairly
+NumPy-specific term, this is called "batching" in deep learning frameworks and
+elsewhere - we plan to change the terminology in NumPy to "batching".
```
Sure, deleted the paragraph and clarified higher up.

rgommers commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+
+- establishing one and way for naming ``inf`` and ``nan`` that is array API
+  compatible.
+- removing cryptic dtype names and establishing canonical names for each dtype.
```
It says that right above, so I'll resolve this comment.

rgommers commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+solving (see `gh-25495 <https://github.com/numpy/numpy/pull/25495>`__).
+
+
+Adapting to the changes
```
Added `& tooling support` to the section title.

rgommers commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+able to support multiple array libraries as easily as possible - are
+significant too. The number of breaking changes needed is fairly limited, and
+the impact of those changes seems modest. Not painless, but smaller than the
+impact of other breaking changes in NumPy 2.0, and a price worth paying.
```
sure, done.

rgommers commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+other key libraries. So those alternatives are not appealing. Given the amount
+of interest in this topic, doing nothing also is not appealing. The "hidden
+namespace" option would be a smaller change to this proposal. We prefer not to
+do that since it leads to duplicate implementations staying around, a more
```
It's not impossible we may need it in the future, but it's not at all clear that we would ever need it. Versioning isn't needed as long as the added API surface and semantics in newer versions of the standard are a superset. Which should be true for the next version at least, and we should try hard to keep it that way. `special` extension isn't a given yet (it could be rejected, and certainly it isn't close to the finish line yet), and otherwise could probably be handled fairly easily in a module-level `__getattr__`.

rgommers commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+- :ref:`NEP37` (never implemented; the ``__array_module__`` idea is basically
+  the same as ``__array_namespace__``)
+- :ref:`NEP47` (implemented with an experimental label in ``numpy.array_api``,
+  will be removed)
```
It seems relevant to me to briefly mention these here - it shows they're related and that a _lot_ of effort went into this topic over the years.

Re NEP 30: I think we don't want that anymore. The whole `__duckarray__` idea revolved around array consumer libraries using `np.asarray` on their inputs, and we're already replacing that by checking for `__array_namespace__` and letting the input array object return a different namespace. I'll defer to @shoyer here though, let's see what he thinks.

rgommers replied:
Stephan replied on the mailing list [here](https://mail.python.org/archives/list/numpy-discussion@python.org/message/Z6AA5CL47NHBNEPTFWYOTSUVSRDGHYPN/), in favor of superseding the other NEPs. For NEP 30 he's the author, and for the authors it's more clear-cut. So I would like to resolve this comment and keep things as they are @seberg.

seberg replied:
Really, I was always OK with keeping it!  Just felt it could be a good option to mark it there and not have to argue about it here.

rgommers commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+default. However, this would be a large effort with some uncertainty about
+technical risks and impact of the change, and no one has taken it on. Given
+that array scalars implement a mostly array-compatible interface, this doesn't
+seem like the highest-prio item regarding array API standard compatibility.
```
https://data-apis.org/array-api/latest/API_specification/array_object.html already contains such a note:

<img width="565" alt="image" src="https://github.com/numpy/numpy/assets/98330/3e7d539c-d8ba-463a-927e-e622beeb7fbb">

rgommers commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+    The main changes proposed in this NEP were presented in the NumPy 2.0
+    Developer Meeting in April 2023 (see `here
+    <https://github.com/numpy/archive/blob/main/2.0_developer_meeting/NumPy_2.0_devmeeting_array_API_adoption.pdf>`__
+    for presentations from that meeting) and given a thumbs up there. The
```
>  (I explicitly did not give a thumbs-up for the type promotion changes in that meeting).

I'm not sure what you mean exactly, since type promotion changes are basically NEP 50. Can you clarify?

rgommers replied:
Not yet, but I will (a few more substantial updates to make). I never considered that a main change, but since it's going to be removed from this NEP I think we're good for this paragraph.

rgommers commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+The ``correction`` name is for clarity ("delta degrees of freedom" is not easy
+to understand) and ``stable`` is complementary to ``kind``, allowing a library
+to reserve the right to change/improve the stable and unstable sorting
+algorithms.
```
adopting that suggestion

rgommers committed:
DOC: second batch of smaller textual changes

[skip actions] [skip azp] [skip cirrus]

rgommers commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+
+There may be other minor changes that don't quite fall in one of the categories
+above. For example, ``numpy.fft`` functions need to preserve precision for
+32-bit input dtypes rather than upcast to ``float64``/``complex128``. And
```
Removed "minor" (I think it is a fairly minor change and expected behavior, but I don't mind - it's just a word). 

Re the more important part (it's gh-25536): probably by casting? IIRC we need to upgrade from `pocketfft` (C) to `pypocketfft` (C++) for native `float32` support. Which is something we also want to do I believe, but it's not done in gh-25536 as far as I can tell.

mhvk replied:
Indeed, gh-25536 just sets the stage by disentangling pocketfft from the wrapper (so, I *think* it should be a lot easier to use the C++ version), and by making it nearly trivial on the numpy side to support other dtypes.  Anyway, really not very important at all for this NEP!

rgommers commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+
+- one or more of the superseded NEPs, or
+- making ``ndarray.__array_namespace__()`` return a hidden namespace with
+  compatible functions,
```
Okay, tweaking text to add "(or even another new public namespace)"

rgommers replied:
done in commit 90df451356

rgommers commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+current place, since it is still useful - it is the best way to test if
+downstream code is actually portable between array libraries. This is a very
+reasonable alternative, however there is a slight preference for taking that
+module and turning it into a standalone package.
```
The whole module is useful for portability testing, and leaving it in is an equally valid choice. So I think this is worth two sentences. It's unrelated to the separate array object.

rgommers replied:
Well it's in the code base and it's used by at least SciPy and scikit-learn, so pretending it never existed at all isn't too helpful I think. It was in with an experimental label, and now we actually have to take action to remove it and deal with the impact of that removal downstream, so I want to describe it as such.

rgommers commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+or data science more seamless.
+
+Finally, for authors of other array libraries as well as tools like Numba,
+array API standard support should save them time. The design rules ([3]_), and
```
That's not what I meant here. What is meant is exactly how it's written "The design rules ...". Numba is always going to want to target the whole NumPy API (at least the whole main namespace and other main modules), so I don't think the suggested edit works.

rgommers replied:
> In that case maybe say: API improvements which align NumPy with the array API will also save time?

Okay, adopted this pretty much verbatim. So I'll resolve the comment.

rgommers commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+  itself has (in cases where there are differences between the two, see for
+  example the `linear algebra design principles <https://data-apis.org/array-api/2022.12/extensions/linear_algebra_functions.html#design-principles>`__
+  and `data-dependent output shapes page <https://data-apis.org/array-api/2022.12/design_topics/data_dependent_output_shapes.html>`__
+  in the standard),
```
adopted the suggestion to move this down, and extended this Benefits section based on another comment.

rgommers commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+the array API standard chose to follow C99 and other libraries in using
+``acos``, ``asin`` etc. rather than ``arccos``, ``arcsin``, etc. NumPy usually
+also follows C99; it is not entirely clear why this naming choice was made many
+years ago.
```
Sure. If I counted correctly, it's 13 aliases in the main namespace and 2 in `numpy.linalg` (xref gh-25086).

rgommers replied:
done in commit 90df451356

rgommers committed:
DOC: adopt suggestions on namespace and alias count/hiding

[skip actions] [skip cirrus] [skip azp]

ngoldbaum commented:
Hi all, I realize that discussion is still ongoing, but since we're running short on time for the NumPy 2.0 RC, I wanted to unblock merging some of the remaining PRs for array API support that did not get any significant discussion so far.

Specifically, I'm planning on merging #25233 (adding `device` keyword arguments in array creation routines and `device` and `to_device` attributes to ndarray) and #25169 (adding a `correction` keyword argument to `np.var` and `np.std`) tomorrow. Please let me know if there are any comments relating to those changes for this NEP that haven't been lodged yet and I will hold off.

seberg reviewed (commented):
Thanks for the updates, looks pretty good.  I simply have a different angle (such as seeing removing `np.array_api` as doing nothing).

There are still a few things that I think are not highlighted (around BC breaks) as much as I would like, but it's fine.
Other than that, I will just point out that I think the alternative isn't terrible in the mailing list:  I don't have a strong enough opinion on new aliases/kwargs to worry, but if anyone really dislikes them, I also think that adding a custom namespace is so cheap that I wouldn't worry if anyone dislikes any of those (or a specific backwards compatibility break).

(I.e. to me most API changes in NumPy are great, but I would be also happy to see them as distinct, because most can be with a namespace.  Maybe, that is misreading the NEP a bit though:  It wants to argue for _not_ doing that, and that is fine.)

rgommers committed:
DOC: extend abstract with the main rationale and benefits

rgommers committed:
add section about which parts of the standard are not adopted

[skip actions] [skip azp] [skip cirrus]

rgommers commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+--------
+
+This NEP proposes adding full support for the 2022.12 version of the array API
+standard in NumPy's main namespace for the 2.0 release.
```
done - please have another look

rgommers commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,524 @@
+2. Making ``cross`` error on size-2 vectors (only size-3 vectors are supported),
+3. Making ``solve`` error on ambiguous input (only accept ``x2`` as vector if ``x2.ndim == 1``),
+4. ``outer`` raises rather than flattens on >1-D inputs,
+5. In-place operators are disallowed when the left-hand side would be promoted.
```
It looks like there is nothing to do here - things are already behaving this way:
```python
>>> # integer dtypes
>>> x = np.ones((2,2), dtype=np.int32)
>>> x += np.ones((2, 2), dtype=np.int64)
>>> x
array([[2, 2],
       [2, 2]], dtype=int32)
>>> x += np.full((2, 2), 2**34, dtype=np.int64)
>>> x
array([[2, 2],
       [2, 2]], dtype=int32)

>>> # floating-point dtypes
>>> x = np.ones((2,2), dtype=np.float32)
>>> x += np.ones((2, 2), dtype=np.int64)
>>> x
array([[2., 2.],
       [2., 2.]], dtype=float32)
>>> x += np.ones((2, 2), dtype=np.float64)
>>> x
array([[3., 3.],
       [3., 3.]], dtype=float32)
>>> x -= x.mean()
>>> x
array([[0., 0.],
       [0., 0.]], dtype=float32)
>>> x -= x.mean().astype(np.float64)
>>> x
array([[0., 0.],
       [0., 0.]], dtype=float32)
```

I think I took this requirement from https://numpy.org/devdocs/reference/array_api.html#type-promotion-differences, but the example given there seems wrong (I checked also against 1.23.5), there is no upcasting. So, false alarm it looks like - sentence can be removed.

rgommers replied:
Now of course there's the opposite issue:
```python
>>> a = np.array(1, dtype=np.int8)
>>> a += np.array(2**12, dtype=np.int16)
>>> a
array(1, dtype=int8)
>>> a + np.array(2**12, dtype=np.int16)
4097
```
It looks like when the right-hand side has a higher-precision dtype, what happens is that it gets cast to the lower-precision dtype silently. Which results in all zeros for values that exceed the max value of the left-hand side dtype. And hence a wrong result. So `x += y` is not equal to `x = x + y` at that point. It would not be strange to completely forbid using a higher-precision dtype on the right-hand side (unrelated to the array API standard though).

seberg replied:
What is to do here is to remove the bullet point!  It would be a huge change and those examples would fail *if you would implement it*.

rgommers replied:
That's what I said: "sentence can be removed.".

That said, this thread is about in-place operators, _not_ about `sum`/`mean` (that's higher up - you just ). I'd say:
1. nothing to do standard-wise
2. not high-impact probably without the `sum`/`mean` change
3. it may be considered a design issue, it's pretty weird that `-=` is so different from the out-of-place version. That's exactly what you said higher up:

> For the result-cast safety: I like it! But I need assurance that it isn't annoyingly disruptive, because if it is I don't buy it is worth doing it, especially skipping a deprecation needs a better reason IMO.

mhvk replied:
I think the array API specification is reasonable here - silent downcasting is very risky, certainly for integers. For the purpose of this NEP, maybe we should punt? It means changing the default casting rule for ufuncs from 'same_kind' to 'safe', which may break a lot of things. Maybe it is something that can be experimented with via some environment variable or so (that controls the default).

rgommers replied:
Makes sense - how about I copy the content here into a new issue, and then we can see if it's possible to slowly deprecate?

mhvk replied:
Yes, sounds good.

rgommers replied:
Ah, looked up the original text, and that clarifies a bit more. There are two requirements:

(1) _An in-place operation must not change the data type or shape of the in-place array as a result of [Type Promotion Rules](https://data-apis.org/array-api/latest/API_specification/type_promotion.html#type-promotion) or [Broadcasting](https://data-apis.org/array-api/latest/API_specification/broadcasting.html#broadcasting)._

(2) _An in-place operation must have the same behavior (including special cases) as its respective binary (i.e., two operand, non-assignment) operation._

I was mainly thinking about (1) - "do not upcast" works perfectly fine. It's (2) that is the issue here. So yes, we do deviate a bit. I'll add it to the "not adopted" section then. (2) is perfectly reasonable as a requirement indeed, but there's no hurry here it looks like.

rgommers replied:
I opened gh-25621 for following up on this topic, and added this requirement to "not adopted" now. Resolving the comment. Thanks for working through it @seberg and @mhvk.

mhvk commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,602 @@
+the name for 2.0 if it had not been part of the array API standard.
+
+
+Parts of the standard that are not adopted
```
Thanks, very helpful!

mhvk reviewed (commented):
Thanks for the updates!  A few more comments in-line.

seberg commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -0,0 +1,602 @@
+to follow (at least at this time). These are:
+
+1. The requirement that an indexing operation must result in an array of the
+   same dtype of the indexed array.
```
Thanks for this section, but not sure where this point comes from?  Sure it returns a different type, but not dtype.
(You "lose" the byte-order storage detail and metadata, but I am sure you don't care about either.)

You can mention that the array-api prefers to not have scalars in some form I guess, for the second part, maybe for that it makes sense to use indexing as an example (it isn't the only place, especially right now).

rgommers committed:
Merge branch 'main' into nep-56-arrayapi

rgommers committed:
Add the in-place "no unsafe casting" item to the not adopted section

Also a few smaller textual changes.

[skip actions] [skip cirrus] [skip azp]

asmeurer commented:
(continuing a discussion that was started elsewhere)

I don't actually see what the issue with scalars is. They duck type as arrays, at least for all the purposes of the standard, so I don't see why they can't be considered as compliant. The text here implies that they aren't but doesn't actually explain why not. In particular [here](https://github.com/numpy/numpy/pull/25542/files#diff-10b0c42a0bdd6097ec8d66b62f823f0d4ee411fb08e020ff447dbbd5770216d0R89) and [here](https://github.com/numpy/numpy/pull/25542/files#diff-10b0c42a0bdd6097ec8d66b62f823f0d4ee411fb08e020ff447dbbd5770216d0R467). The second note just seems wrong. The only reason that scalars don't work for "an array of the same data type as the indexed array" is if you don't consider scalars to be an array. But this argument is circular, because why isn't a scalar an array, for the purposes of the standard? 

The only real "not array" thing about scalars is that they are immutable, but mutability is optional in the standard. That and the fact that they aren't called "arrays" in the NumPy docs, but pytorch doesn't call its arrays "arrays" either.

asmeurer commented:
And just to be clear, the standard doesn't require all arrays from a given library to have the same type. This is even discussed explicitly https://data-apis.org/array-api/latest/design_topics/static_typing.html

> Also note that this standard does not require that input and output array types are the same (they’re expected to be defined in the same library though).

mhvk commented:
Good point. From a *numpy* user perspective, the main annoyance I have with scalars is that they share most but not all methods with arrays (e.g., they cannot be `.view`d). But that's pretty irrelevant for the discussion here.

rgommers commented:
> I don't actually see what the issue with scalars is. They duck type as arrays, at least for all the purposes of the standard

Good points. In principle yes, you are right - since it's not required that the returned 0-D array is of the same type as the input array and there's no `isinstance` check, duck typing should fulfill all the requirements here.



> From a _numpy_ user perspective, the main annoyance I have with scalars is that they share most but not all methods with arrays

Here's one such paper cut:
```python
>>> x_0d.mT
...
ValueError: matrix transpose with ndim < 2 is undefined

>>> np.float32(3.5).mT
...
AttributeError: 'numpy.float32' object has no attribute 'mT'
```

The standard doesn't require exceptions of a particular type, so this is technically still fully compliant. It's just annoyances. 

For this PR, I plan to rework the text to not state that scalars are non-compliant and an explanatory note saying that while there are minor differences, they duck type well enough to be considered the same as 0-D arrays from the perspective of the standard.

rgommers committed:
Merge branch 'main' into nep-56-arrayapi

rgommers committed:
Merge branch 'main' into nep-56-arrayapi

rgommers committed:
Merge branch 'main' into nep-56-arrayapi

rgommers committed:
NEP: update content mentioning array scalars

Addresses review feedback. Array scalars aren't actually incompatible
with the design of the standard.

rgommers committed:
NEP: improve backwards compatibility section and describe `copy=` better

As requested in PR review, better delineate the different types of
changes in the backwards compatibility section, and add assessments
of what the impact of changes is.

In addition, expect the description of the `copy=` keyword semantics
changes.

[skip cirrus] [skip actions] [skip azp]

rgommers commented:
I pushed two new commits. The first one changes how array scalars are described. The second one adds more detail to the backwards compatibility section (including impact assessment per category of change) and expands on the `copy=` keyword.

rgommers commented:
I'll note that pretty much all PRs implementing this NEP (except for the copy keyword one) have been merged for quite a while now, so the impact assessment is based on at least a month of usage by libraries and end users that test against NumPy main. Over the course of 2.0 development we've reverted several backwards incompatible changes that turned out to be disruptive; nothing that is in the backwards compatibility section now falls into that category I believe.

mhvk commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -272,15 +288,31 @@ same name. We may deprecate ``np.trace`` and ``np.diagonal`` to resolve it, but
+lower overhead than* ``np.asarray(...)``. *This was solved though, and*
+``np.asarray(...)`` *is idiomatic NumPy usage.*
+
+**Other changes**: there may be a few other changes that don't quite fall in
```
This has now happened - gh-25711

mhvk commented on doc/neps/nep-0056-array-api-main-namespace.rst:
```diff
@@ -272,15 +288,31 @@ same name. We may deprecate ``np.trace`` and ``np.diagonal`` to resolve it, but
+gufunc implementation in (`gh-25336
+<https://github.com/numpy/numpy/pull/25336>`__) . Also in ``numpy.fft``,
+there's an issue with the ``s``/``axes`` argument in n-D transforms that needs
+solving (see `gh-25495 <https://github.com/numpy/numpy/pull/25495>`__).
```
gh-25495 has also been merged.

rgommers replied:
Yes, great that these have been merged. I'm not sure that should change the text here though? Ideally it reads okay whether or not linked PRs have been merged or not.

mhvk replied:
It just seemed the text would read more easily if adjusted. (and the first link points to an unrelated PR...)

rgommers replied:
Okay, you are right. I rewrote this section now and make it explicitly about `numpy.fft` rather than putting it under "other changes".

rgommers committed:
Merge branch 'main' into nep-56-arrayapi

rgommers committed:
NEP: update description of `fft` changes

[skip actions] [skip cirrus] [skip azp]

ngoldbaum commented:
There's still one outstanding PR: https://github.com/numpy/numpy/pull/25802 and it looks like discussion around the `copy` argument for `asarray` has settled down here. I'd like to merge the PR implementing it tomorrow. Let me know if you'd like me to hold off.

rgommers committed:
Merge branch 'main' into nep-56-arrayapi

[skip azp] [skip cirrus] [skip actions]

rgommers commented:
I think it's time (or well overdue) to merge this PR with Draft status, so we have a rendered version. Discussion seems to have settled down and all comments are addressed. In case someone wants to return to a comment that they think wasn't addressed well enough yet or if we have a hiccup with one of the recent PRs implementing support for this NEP, we can always open a follow-up PR.

charris commented:
Thanks Ralf and thanks to all who reviewed and offered comments.
